lib/imagec/imagec.go

```diff
@@ -569,16 +682,325 @@ func (ic *ImageC) PullImage() error {
 	if err != nil {
-		return err
+		return fmt.Errorf("Could not retrieve image id from repository cache using reference: %s", err)
 	}
-	ic.ImageLayers = layers
 
-	err = ldm.DownloadLayers(ctx, ic)
+	pusher := &ic.Pusher
+
+	// get the leaf layerID from repo cache using the image id
+	layerID := cache.RepositoryCache().GetLayerID(id)
+
+	// get the layer (ImageWithMeta) from the layer cache using the layer id
+	layer, err := LayerCache().Get(layerID)
 	if err != nil {
-		return err
+		return fmt.Errorf("Unable to get top layer id for image %s", id)
+	}
+
+	// Build a layers history slice and add layers to stream map
+	var layersHistory []*ImageWithMeta
+	layersHistory = append(layersHistory, layer)
+
+	// use the leaf layer to walk the chain of layers down to the base parent (scratch)
+	for {
+		pusher.streamMap[layerID] = &ArchiveStream{
+			layerID:       layerID,
+			parentLayerID: layer.Image.Parent,
+		}
+
+		log.Infof("Image data for layer %s = %#v", layerID, *layer.Image)
+
+		// Check for scratch ID
+		if layer.Image.Parent == storage.Scratch.ID {
+			break
+		}
+
+		// set the layer to the parent layer
+		layerID = layer.Image.Parent
+		layer, err = LayerCache().Get(layerID)
+		if err != nil {
+			return fmt.Errorf("Unable to get metadata for layer %s", layerID)
+		}
+
+		layersHistory = append(layersHistory, layer)
+	}
+
+	log.Infof("streamMap = %#v", pusher.streamMap)
+
+	// Create a docker image from our (VIC's) image config
+	_, configDigest, configSize, err := ic.dockerImageFromVicImage(layersHistory)
+	if err != nil {
+		return fmt.Errorf("Could not build docker image from VIC image history: %s", err.Error())
+	}
+
+	// calculate image ID
+	log.Infof("Image ID: sha256:%s", configDigest)
+	digest := "sha256:" + digest.Digest(configDigest)
+
+	// build out PushManifest with all generated components
+	pusher.PushManifest = schema2.Manifest{
+		Versioned: schema2.SchemaVersion,
+		Config: distribution.Descriptor{
+			MediaType: schema2.MediaTypeImageConfig,
+			Size:      configSize,
+			Digest:    digest,
+		},
+	}
+
+	log.Infof("schema 2 manifest: %#v", pusher.PushManifest)
+
+	return nil
+}
+
+// FinalizeManifest builds the layer information in the manifest.  This information cannot
+// be determined till the layer tar streams were read and pushed.
+func (ic *ImageC) FinalizeManifest() error {
+	defer trace.End(trace.Begin(""))
+
+	pusher := ic.Pusher
+
+	var layers []distribution.Descriptor
+	for _, stream := range pusher.streamMap {
+		var layer distribution.Descriptor
+
+		log.Infof("Finalizing manifest, stream: %#v", stream)
+		layer.Digest = digest.Digest(stream.digest)
+		layer.Size = stream.size
+		layer.MediaType = schema2.MediaTypeLayer
+
+		layers = append(layers, layer)
 	}
 
+	pusher.PushManifest.Layers = layers
+
+	log.Infof("Final manifest = %#v", pusher.PushManifest)
+
 	return nil
 }
+
+func (ic *ImageC) pushManifest(ctx context.Context) error {
+	if ic.ImageManifestSchema2 != nil {
```

Comment:
This might be a problem. Where is this `ImageManifestSchema2` assigned? In my local testing, it is always nil. Since we only construct image manifest following schema2, shall we remove this check?